Client/Frontend/Home/ActivityStreamPanel.swift

```diff
@@ -0,0 +1,337 @@
+                    return 1
+                } else {
+                    return 0
+                }
+            case .history:
+                 return self.history.count
+        }
+    }
+
+    func tableView(tableView: UITableView, cellForRowAtIndexPath indexPath: NSIndexPath) -> UITableViewCell {
+        let identifier = Section(indexPath.section).cellIdentifier
+        let cell = tableView.dequeueReusableCellWithIdentifier(identifier, forIndexPath: indexPath)
+
+        switch Section(indexPath.section) {
+            case .topSites:
+                return configureTopSitesCell(cell, forIndexPath: indexPath)
+            default:
+                return configureHistoryItemCell(cell, forIndexPath: indexPath)
+        }
+    }
+
+    func configureTopSitesCell(cell: UITableViewCell, forIndexPath indexPath: NSIndexPath) -> UITableViewCell {
+        let topSiteCell = cell as! ASHorizontalScrollCell
+        // The topSiteCell needs a refrence to the tableview because it needs to alert the tableView to relayout once topsites finishes loading.
+        topSiteCell.parentTableView = self.tableView
+        topSiteCell.setDelegate(self.topSiteHandler)
+        return cell
+    }
+
+    func configureHistoryItemCell(cell: UITableViewCell, forIndexPath indexPath: NSIndexPath) -> UITableViewCell {
+        let simpleHighlightCell = cell as! SimpleHighlightCell
+        let site = history[indexPath.row]
+        simpleHighlightCell.configureSimpleHighlightCell(site)
+        return simpleHighlightCell
+    }
+
+    func showSiteWithURL(url: NSURL) {
+        let visitType = VisitType.Bookmark
+        homePanelDelegate?.homePanel(self, didSelectURL: url, visitType: visitType)
+    }
+
+    func tableView(tableView: UITableView, didSelectRowAtIndexPath indexPath: NSIndexPath) {
+        switch Section(indexPath.section) {
+        case .history:
+            let site = self.history[indexPath.row]
+            showSiteWithURL(NSURL(string:site.url)!)
+        default:
+            return
+        }
+    }
+}
+
+// MARK: - Data Management
+extension ActivityStreamPanel {
+
+    func notificationReceived(notification: NSNotification) {
+        switch notification.name {
+        case NotificationProfileDidFinishSyncing:
+            // Only reload top sites if there the cache is dirty since the finish syncing
+            // notification is fired everytime the user re-enters the app from the background.
+            self.profile.history.areTopSitesDirty(withLimit: ASPanelUX.topSitesCacheSize) >>== { dirty in
+                if dirty {
+                    self.refreshTopSites(ASPanelUX.topSitesCacheSize)
+                }
+            }
+        case NotificationFirefoxAccountChanged, NotificationPrivateDataClearedHistory, NotificationDynamicFontChanged:
+            self.refreshTopSites(ASPanelUX.topSitesCacheSize)
+        default:
+            // no need to do anything at all
+            print("Dont have this notification type")
```

Comment:
Not sure if this is temporary or not but when logging use an instance of XCLogger from Logger.swift instead of print.